Show and switch groups

ElbertTse opened the pull request:
Implements group switching which changes the displayed images, invite link, main-content title, and adds font weight to selected group. Also adds a fallback image if given a bad URL for an image.

ElbertTse committed:
Made rough plan of attack

ElbertTse committed:
Made GroupContext

ElbertTse committed:
Made dummy groups list

ElbertTse committed:
Added images to dummy group list

ElbertTse committed:
Renamed groups to Groups and set up in mainpage

ElbertTse committed:
Added states for groups, currentGroup, and photos

ElbertTse committed:
Is switch groups but lags since we arent waiting for setState to finish

ElbertTse committed:
Fixed delay for switching cards and changed Card children proptype to node

ElbertTse committed:
Moved group context up tree, created new group context object

ElbertTse committed:
Changing title with group change

ElbertTse committed:
Added error handling for group title header

suneettipirneni committed:
Use dispatch context data flow

ElbertTse committed:
Added guards for showing empty group list

ElbertTse committed:
Showing and switching groups in sidebar

ElbertTse committed:
Sidebar group items reflect actual groups

ElbertTse committed:
Changing invite link

ElbertTse committed:
Changing group title with selection

ElbertTse committed:
Changed value to placeholder in invite area

ElbertTse committed:
Fixed elilipsis flipping

ElbertTse committed:
Chaging font weight of selected group

ElbertTse committed:
Added fallback image

ElbertTse committed:
Removed TODOs and moved photo array to groups

ElbertTse committed:
Fixing deepsource issues

ElbertTse committed:
Cleaned up logic in MainPage

ctcuff commented on src/components/dashboard/InviteArea.jsx:
```diff
@@ -46,7 +47,7 @@ function InviteArea({ inviteCode }) {
         <TextInput
           className="invite-link-input"
-          value={inviteCode ? inviteLink : ''}
+          placeHolder={inviteCode ? inviteLink : ''}
```
Any way you could use `value` here instead of `placeHolder`? Just nitpicking but it'd still be nice to be able to select the text from the input manually.

ElbertTse replied:
So for whatever reason, the text wouldn't change on switches if we placed the inviteLink in value, but would work if I placed it in placeHolder. I can give it another try and comment back on this thread.

ElbertTse replied:
Double checked, using value wont let the inviteLink change but using placeHolder does.

ctcuff commented on src/components/dashboard/PhotoThumbnail.jsx:
```diff
@@ -22,7 +22,11 @@ function PhotoThumbnail({ src }) {
+        className="photo-thumbnail"
+        src={src}
+        fallback="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3hWNV2E6aTe_edfDRBjSKo4KvAWfzaXxHzw&usqp=CAU"
+      />
```
It may be better to download a placeholder image, put that in our `assets` folder, and use that. That way the browser doesn't have to try to download a placeholder image from a website (since even that image might not be available). Also not sure if we should go with that specific image but considering I haven't found a good one yet we can come back to it later.

ElbertTse replied:
Ill go ahead and download it the image I'm displaying now but I'll send some different fall back image options before I resolve the thread.

ctcuff commented on src/contexts/GroupsContextDispatch.jsx:
```diff
@@ -0,0 +1,19 @@
+}
+
+export default GroupsDispatchContext;
+export { groupReducer };
```
Another nitpick but this could be simplified to
```js
export { GroupsDispatchContext as default, groupReducer }
```

ElbertTse replied:
Ill make that change.

ctcuff commented on src/components/dashboard/GroupList.jsx:
```diff
@@ -18,12 +18,18 @@ function GroupList({ groups, onGroupClick }) {
-    <List.Item onClick={() => onGroupClick(groups[index])} title={title}>
+  const renderTitle = (title, ind) => (
+    <div className={ind === index ? 'selected' : ''}>
+      {title}
```
Noticed that the selected group is bold now 💯, but it's missing the ellipses styling when the group name is too long to be displayed.

ElbertTse replied:
Good catch, I'll fix that.

ctcuff commented on src/components/DisplayBox.jsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
```
Doesn't look like this file is used anywhere? If not you can go ahead and remove it.

ElbertTse replied:
Removed it

ctcuff commented on src/components/dashboard/GroupList.jsx:
```diff
@@ -3,7 +3,7 @@ import React from 'react';
 import PropTypes from 'prop-types';
 
-function GroupList({ groups, onGroupClick }) {
+function GroupList({ groups, index, onGroupClick }) {
```
I think `activeIndex` is probably a more clear prop name here.

ElbertTse replied:
Changed the name

ctcuff commented on src/pages/MainPage.jsx:
```diff
@@ -9,75 +9,105 @@ import Sidebar from '../components/dashboard/Sidebar.jsx';
   useEffect(() => {
     const token = localStorage.getItem('token');
     const id = localStorage.getItem('id');
     getUser(token, id);
```
Since `getUser` is async, there should be an `await`, that way we don't try to keep loading the main page if `getUser` fails

ctcuff reviewed (changes requested):
Just a few changes/suggestions but it's nice to finally see the main page working!

ElbertTse committed:
Made GroupContextDispatch export more concise

ElbertTse committed:
Renamed index prop in GroupList

ElbertTse committed:
using value prop in InviteArea works now

ElbertTse committed:
Fixed ellipse issue

ElbertTse committed:
Added Cam's cleanup

ctcuff committed:
Formatted with prettier and simplified list title render logic

ctcuff committed:
Minor cleanup

- The "Copy to clipboard button" no longer copies empty text
- Changed Sidebar to check for length instead of undefined
- Added "https" to copied invite link

ctcuff committed:
Merge branch 'main' of https://github.com/Imageus-OSS/client into design/show-and-switch-groups

ctcuff reviewed (approved):
LGTM :+1:, added a few small commits, mostly just formatted with prettier and simplified how the title renders an active item.

ElbertTse committed:
Changed comment to restart aws action
